[ready for review] Batched upper triangular, lower triangular

vishwakftw opened the pull request:
Changelog:

- Implements `triu` and `tril` for batches of 2D tensors.
- Remove TH/THC binding for `tril`
- Fix CUDA implementation
- Update docstrings for tril and triu.
- Remove mask-based `triu` and `tril` in cholesky forward and backward.
- Remove batched tril in torch.distributions.utils

Test plan:
- Add tests for tril and triu for CPU and CUDA.

Fixes #15016, fixes #15226 and closes #14071

Acknowledgements:
- Thanks to @t-vi whose implementation I used as a reference.

vishwakftw committed:
Upper/Lower triangular for batch of matrices

vishwakftw commented:
I've tried to debug this error, but to no avail. Posting it below:
```
Dec 15 07:42:30 /var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDAApplyUtils.cuh(331): error: no instance of overloaded function "at::native::BatchTensorTriOp<T, upper>::operator() [with T=uint8_t, upper=false]" matches the argument list
Dec 15 07:42:30             argument types are: (uint8_t, uint8_t)
Dec 15 07:42:30             object type is: const at::native::BatchTensorTriOp<uint8_t, false>
Dec 15 07:42:30           detected during:
Dec 15 07:42:30             instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, 0, Offset>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offset, Offset) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, Offset=const unsigned int]" 
Dec 15 07:42:30 (310): here
Dec 15 07:42:30             instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, remaining_steps, Offsets...>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offsets..., Offsets...) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, remaining_steps=1, Offsets=<>]" 
Dec 15 07:42:30 (369): here
Dec 15 07:42:30             instantiation of "void at::cuda::<unnamed>::kernelPointwiseApply2<Op,scalar1,scalar2,IndexType,ADims,BDims,step>(at::cuda::detail::TensorInfo<scalar1, IndexType>, at::cuda::detail::TensorInfo<scalar2, IndexType>, IndexType, Op) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, step=1]" 
Dec 15 07:42:30 (888): here
Dec 15 07:42:30             instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,step,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, step=1, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
Dec 15 07:42:30 (940): here
Dec 15 07:42:30             instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
Dec 15 07:42:30 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(430): here
Dec 15 07:42:30             instantiation of "void at::native::apply_triu_tril<scalar_t,inplace,upper>(at::Tensor &, const at::Tensor &, int64_t) [with scalar_t=uint8_t, inplace=true, upper=false]" 
Dec 15 07:42:30 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(437): here
```

zou3519 commented:
@vishwakftw do you get the same error message on a local build?

vishwakftw commented:
Yes, I do.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -376,6 +376,93 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+    return upper ? (col - row >= k) : (col - row <= k);
+  }
+
+  __device__ __forceinline__ void operator()(T* result, T* self) {
```
Does changing this to ` __device__ __forceinline__ void operator()(T& result, T& self) {`

(and the corresponding operator()(T* v) to take a reference) help?

zou3519 replied:
The operator passed to CUDA_tensor_apply2 should take in references, not pointers, so I think this should fix your compilation error: https://github.com/pytorch/pytorch/blob/bb9b7de831d25ab40d64aca6069d8166ee5d92f7/aten/src/ATen/native/cuda/Activation.cu#L21-L29

vishwakftw replied:
Thank you, I’ll check if this works out locally.

vishwakftw replied:
No, this doesn't work unfortunately. The leading error in this case is:

```
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/cuda/CUDAApplyUtils.cuh(331): error: no instance of overloaded function "at::native::BatchTensorTriOp<T, upper>::operator() [with T=uint8_t, upper=false]" matches the argument list and object (the object has type qualifiers that prevent a match)
            argument types are: (uint8_t, uint8_t)
            object type is: const at::native::BatchTensorTriOp<uint8_t, false>
          detected during:
            instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, 0, Offset>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offset, Offset) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, Offset=const unsigned int]" 
(310): here
            instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, remaining_steps, Offsets...>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offsets..., Offsets...) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, remaining_steps=1, Offsets=<>]" 
(369): here
            instantiation of "void at::cuda::<unnamed>::kernelPointwiseApply2<Op,scalar1,scalar2,IndexType,ADims,BDims,step>(at::cuda::detail::TensorInfo<scalar1, IndexType>, at::cuda::detail::TensorInfo<scalar2, IndexType>, IndexType, Op) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, step=1]" 
(888): here
            instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,step,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, step=1, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
(940): here
            instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(430): here
            instantiation of "void at::native::apply_triu_tril<scalar_t,inplace,upper>(at::Tensor &, const at::Tensor &, int64_t) [with scalar_t=uint8_t, inplace=true, upper=false]" 
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(437): here
```

This reminded me that I had already tried this earlier, and this didn't work. In fact, this was how @t-vi had kept it in his original PR.

t-vi replied:
I don't know much about it except that my patch worked at the time of submission, but so the argument types it tries to call with are uint8_t, uint8_t and so if you declare that the first needs to be a const, it would not work.

vishwakftw replied:
I circumvented this issue by using thrust::tabulate. Thank you @zou3519 @t-vi for your suggestions, nonetheless. :)

vishwakftw committed:
Use thrust for CUDA tril/triu, update doc strings, add new tests

vishwakftw committed:
Use a custom kernel instead of thrust for the CUDA implementation, optimize CPU implementation

- The thrust implementation seemed to be incredibly slow
- The CPU implementation was bottlenecked by a clone() op
- Add test cases for non-square matrices, and an addition based test

vishwakftw commented:
@zou3519 is this good to go?

vishwakftw commented:
I did some benchmarking of the PR versus current master for the triu / tril ops on CPU and CUDA. Below are links to the results:
triu CPU : https://gist.github.com/vishwakftw/83b11cdcb18db19365d6c756cf214665
tril CPU : https://gist.github.com/vishwakftw/87f8d55b32c8586599fe6a2c25a186e1
triu CUDA : https://gist.github.com/vishwakftw/403e4d9d5d404c64a6456d5c4a5cbfe6
tril CUDA : https://gist.github.com/vishwakftw/7cdba13940c04b205e721e27f3b3e7f9

vishwakftw committed:
Remove batch_tril in torch.distributions.utils

zou3519 commented:
@vishwakftw I'll take a look later today

vishwakftw commented:
Thank you, much obliged.

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -5090,19 +5090,19 @@ def parse_kwargs(desc):
-            [-0.2447,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
-            [ 0.4333,  0.3146,  0.0000,  0.0000,  0.0000,  0.0000],
-            [-0.9888,  1.0679, -1.3337,  0.0000,  0.0000,  0.0000]])
+    >>> torch.triu(b, diagonal=1)
```
Nice catch!

vishwakftw replied:
Thank you.

zou3519 commented on test/test_torch.py:
```diff
@@ -3765,7 +3765,6 @@ def test_tril(self):
     def test_trilu_indices(self):
```
There's a test right above this named "test_tril": should we remove this now?

zou3519 commented on test/test_torch.py:
```diff
@@ -3776,12 +3775,31 @@ def test_trilu_indices(self):
+            self.assertEqual(x, x.triu(diagonal) + x.tril(diagonal - 1))
+
+        diagonals = [-2, -1, 0, 1, 2]
+        shapes = [(3, 3), (5, 3, 3), (7, 5, 3, 3),  # square matrices
```
Can you test the following (edge) cases as well:
- 0-dim tensor (size `[]`)
- Tensor with 0 in one dimension (size `[0, 3, 3]` and/or `[3, 0]`)
- Tensor with 1 in one dimension (size `[1, 3]`, `[3, 1]`)

vishwakftw replied:
Thanks for the suggestions. These have been added to the suite as well.

zou3519 commented on test/test_torch.py:
```diff
@@ -3776,12 +3775,31 @@ def test_trilu_indices(self):
-        torch.triu(x, out=res2)
-        self.assertEqual(res1, res2, 0)
+    @staticmethod
+    def _test_triu_tril(self, cast):
```
Is there another triu test somewhere? the original one is... really weird

vishwakftw replied:
No, I removed the triu test. I seem to have missed the tril test while fixing a merge conflict

zou3519 commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -2083,55 +2083,6 @@
 ]]
-[[
-  name: _th_tril
-  cname: tril
```
Does this mean we can delete the old TH code now? i.e. `THTensor_(tril)`

vishwakftw replied:
I have removed tril. triu is required for a Lapack function, and both tril and triu for THC are also required for certain Magma ops.

zou3519 reviewed (commented):
CPU code looks fine, still reading the CUDA kernel

vishwakftw committed:
Revert "Remove batch_tril in torch.distributions.utils"

This reverts commit e907c7414e64f00199b546154c2bcec8de114ad1.

vishwakftw committed:
CR.1
- Add edge cases
- Remove redundant tests
- Fix for non-contiguous case
- Pop batch_tril in torch.distributions.utils
- Remove tril in TH only (cannot remove triu due to requirement in THTensorLapack, cannot remove either tril or triu due to dependencies in THCTensorMathMagma.cu)

vishwakftw commented:
CUDA tests are failing with this error: `RuntimeError: cuda runtime error (9) : invalid configuration argument at /var/lib/jenkins/workspace/aten/src/THC/THCTensorMathCompareT.cuh:69`

vishwakftw committed:
Add a proper tril/triu test and fix empty tensor case for CUDA. Also enforce contiguity implicitly.

vishwakftw commented:
All errors should hopefully be fixed by now.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -15,6 +15,8 @@
 #include <magma_types.h>
 #endif
 
+#include <stdio.h>
```
is this include intentional?

vishwakftw replied:
Oh no. Sorry about that, I used it for debugging purposes. Will remove this in the commit after your code review. Thank you for pointing it out.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,133 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  if (self.numel() == 0) {
+    return self;
+  }
+  if (!self.is_contiguous()) self = self.contiguous();
```
Why is this necessary? I don't see this in the legacy TH code

vishwakftw replied:
Tensors defined via the expand operation are non-contiguous and share memory. The current  implementation failed due to invalid memory accesses which caused the distributions tests to fail earlier.

I think this point was not addressed in the TH, but I might be wrong.

zou3519 replied:
Interesting. I think the problem is the code doesn't support tensors with 0 in their strides -- this code as well as the TH code should work for all other tensors. Can you do a `self.contiguous` call only when the input tensor has a 0 in its stride? In general, `tensor.contiguous()` calls get really expensive the larger a tensor is / the more dimensions it has so we try to avoid these as much as possible.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,133 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  return result;
+}
+
+Tensor triu(const Tensor& self, int64_t k) {
```
nit: I'm pretty sure the gradients are correct, but for correctness can you add an extra entry to https://github.com/pytorch/pytorch/blob/c86cd9e530e7ebee033c5d86465ece756fe6e1ff/test/common_methods_invocations.py#L570-L573 to make an autograd test that tests batch triu / tril?

vishwakftw replied:
Sure, will add this too. Thank you again.

zou3519 commented on test/test_torch.py:
```diff
@@ -3776,12 +3768,55 @@ def test_trilu_indices(self):
-        torch.triu(x, out=res2)
-        self.assertEqual(res1, res2, 0)
+    @staticmethod
+    def _test_triu_tril(self, cast):
```
Could you also add some tests for non-contiguous-inputs to `torch.triu, tensor.triu_, torch.tril, and tensor.tril_`?

vishwakftw replied:
Yes, I will get this done. Thank you for the suggestions.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,133 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  if (self.numel() == 0) {
+    return result;
+  }
+  Tensor self_c = self.is_contiguous() ? self : self.contiguous();
```
Is this contiguous call necessary?

vishwakftw replied:
I’ll look into this, thank you.

vishwakftw replied:
I'll look into this, thanks.

zou3519 reviewed (commented):
Kernels look good, two comments:

- I'm not sure about the condition for adding `.contiguous()` (let's try to avoid these as much as possible) because the kernels look like they should handle most non-contiguous input.
- let's test some non-contiguous inputs as well, I'm surprised we didn't have tests for those before

vishwakftw committed:
CR.2

- Add tests for non-contiguous inputs and inplace version
- Restrict .contiguous calls, use only when strides have 0s in them
- Elaborate on logic used for mask in kernel
- Remove unnecessary header

zou3519 commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -41,6 +41,14 @@ static inline int64_t matrixStride(const Tensor& batched_matrices) {
 
+// Checks if a given tensor has 0s in its stride
+static inline bool checkZeroStride(const Tensor& tensor) {
+  for (auto& stride : tensor.strides()) {
```
nit: `const auto&`

vishwakftw replied:
Thank you, sorry about that.

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,133 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+}
+
+Tensor& tril_cpu_out(Tensor &result, const Tensor& self, int64_t k) {
+  result = at::empty_like(self);
```
again I'm not sure if this assigns the Tensor or uses an overloaded `operator=` to instead do a `result.copy_(at::empty_like(self))`.

`torch.tril(input, diagonal=0, out=result)` should `resize_` `result` to be the correct size so maybe this line should be a result.resize_as_(self)?

vishwakftw replied:
I thought so too. I wasn't sure if undefined tensors could be resized appropriately, which is why I didn't do this.

zou3519 replied:
I don't think the user is allowed to pass in undefined tensors, so it is fine not to worry about them here

vishwakftw replied:
Sure, I'll get this done.

zou3519 commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -371,6 +371,144 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  return result;
+}
+
+Tensor& triu_cuda_(Tensor &self, int64_t k) {
```
Is it possible to deduplicated `triu_cuda_` and `triu_cuda_out`?
`triu_cuda_` is almost `triu_cuda_out(self, self, k)` (that's what it was written as in the TH code I think)

vishwakftw replied:
I'll look into this, thank you.

vishwakftw commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -448,7 +448,7 @@ Tensor& tril_cpu_(Tensor &self, int64_t k) {
     return self;
   }
-  if (!self.is_contiguous()) self = self.contiguous();
+  if (checkZeroStride(self)) self = self.contiguous();
```
I suppose it changes the tensor. This was also tested in the new tests.

However, I am not sure if copying can modify the geometry of the underlying tensor.

vishwakftw replied:
edit: copying doesn't affect the geometry, so I guess this changes the tensor directly.

zou3519 replied:
https://github.com/pytorch/pytorch/blob/c1643ec551ea5a5323f4de4df5805089f176bc54/aten/src/ATen/core/Tensor.h#L85-L114 so it doesn't copy the tensor data. Can you check if the in-place op actually works, for this case, then? If the user passes in a tensor `self` in python, we'll be modifying self.contiguous() in-place, so no updates to `self` should happen (which is not what we want)

vishwakftw replied:
In the in-place case, if there are 0s in the stride, we make them contiguous.

I've tested this behavior here:
https://github.com/pytorch/pytorch/blob/c92d1544b7512616c2d3b18ae57eea471556a339/test/test_torch.py#L3817-L3829

vishwakftw replied:
Sorry if I missed something, is this what you think wasn't happening?

zou3519 replied:
I see, thanks for pointing ting out that test. I'm definitely missing something (the code works), let me think about this more

vishwakftw committed:
CR.3

- auto& stride -> const auto& stride
- deduplicate code in tri*_cuda
- resize result in tri*_out instead of creating a empty tensor

vishwakftw commented:
@zou3519 is this good to go?

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -371,6 +371,126 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  int64_t min_stride = row_is_max_stride ? self_col_stride : self_row_stride; \
+  dim3 dim_block = cuda::getApplyBlock(); \
+  dim3 dim_grid; \
+  AT_CHECK(cuda::getApplyGrid(N, dim_grid, SELF.get_device()), "Unable to get dim grid");
```
Typical way of handling batched computations is to add grid dimension (dim_grid.y for example) to handle batch elements. That way, you have less index math in the kernel that is good both for readability and for performance.

vishwakftw replied:
Would the following be appropriate?:
```
dim3 dim_grid((N + dim_block.x - 1) / (dim_block.x * batchCount(SELF)), batchCount(SELF));
```

`batchCount(SELF)` is a function to return the number of matrices batched. This would allow me to use `gridDim.y` as a batch index, as advised.

vishwakftw commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -371,6 +371,126 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+        result.data<scalar_t>(), self_c.data<scalar_t>(), k, N,
+        res_batch_stride, self_batch_stride, res_row_stride, res_col_stride,
+        max_stride, min_stride, row_is_max_stride);
+  });
```
Sorry about that, will incorporate.

vishwakftw commented on test/test_torch.py:
```diff
@@ -3776,12 +3768,85 @@ def test_trilu_indices(self):
+                # non-contiguous and expanded tensors test
+                if not (0 in shape or 1 in shape):
+                    # non-contiguous tensors
+                    x_nc = x.clone().transpose(-2, -1)
```
Sure, thank you.

vishwakftw commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -371,6 +371,126 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  result[res_offset] = mask ? self[linear_idx] : scalar_t(0);
+}
+
+#define DECLARE_TRIL_TRIU_VARIABLES(SELF, RESULT) \
```
I originally intended on doing that, but since there were a lot of arguments, I did this instead.

vishwakftw commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -371,6 +371,126 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
   }
 }
 
+#define HANDLE_EQUATION_CASE(MAX1, MAX2, MAX3) \
```
Oh yes, you're right. How do you suggest I convert linear_index to offset?

ngimel replied:
You don't need to, you can calculate col and row based on linear_index, and you won't even need conditionals for that (col = linear_index % size[-1], row = linear_index/size[-1] if your linear index is an index into an individual matrix, batch should be handled separately). In THC, they were forced to use offset, but you don't have this limitation.

vishwakftw commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,137 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  if (self.numel() == 0) {
+    return self;
+  }
+  if (checkZeroStride(self)) self = self.contiguous();
```
I'm not sure what to do here, perhaps it would be better to consult a PyTorch core-dev about this?

zou3519 commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -375,4 +367,137 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  int64_t b;
+
+  #pragma omp parallel for private(b)
+  for (b = 0; b < batchsize; b++) {
```
Yes, you are correct, thanks for catching that. I don't think there's an easy way to iterate over the batches. One way we could resolve the problem is call `tensor.contiguous()` if the batch dimensions are non-contiguous and then the rest of the code should work.

The alternative approach would be to convert a linear index (that ranges from 0 to `tensor.numel()`) to an index vector, then do a dot product with the strides to get the memory location.

vishwakftw replied:
I think the former approach would be relatively cheaper, given that the dot product would have to be computed multiple times.

While analyzing this problem, I also thought of a mask based solution. While I am not sure if it would be faster, I am certain that it would retain the (non-)contiguity of the input tensor. We can create a mask using `apply_triu_tril_single`, and then expand it to the required size, after which `at::where` can be used to obtain the result.

zou3519 replied:
I agree that the first approach is cheaper and would probably be less overhead for non-batched triu/tril. Does the same problem exist on the CUDA side? It might be a little strange if the CUDA operator doesn't make the tensor contiguous in some cases when the CPU operator does

ngimel replied:
This is not properly handled on the cuda side at all https://github.com/pytorch/pytorch/pull/15257/files#diff-5b3ae8ff75409dce3f752955b36b9359R433
1) For a batched case, all batch dimensions are assumed to be contiguous (batch_stride = tensor.stride(-3))
2) For a non-batched case batch stride is calculated as rows*cols, which is not right, but probably benign.

vishwakftw replied:
I am sorry, I will fix this case as soon as I can, and keep you posted.

As suggested, I will make the tensor contiguous only if the batch dims are non-contiguous.

vishwakftw committed:
CR.4

- Clean up the kernel
- Deduplicate the kernel launches by introducing a template function
- Add error checking after kernel launch
- If batch is not contiguous, then make the tensor contiguous
- Introduce other dim transpose test (other than the matrix dimensions)

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batched-tril-triu

vishwakftw committed:
Remove tril from THC, as it is no longer required

vishwakftw committed:
Remove mask based triu / tril for btriunpack

vishwakftw commented:
@ngimel @zou3519 I have made changes as recommended. Could you please take a look?

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -461,6 +461,81 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+template <typename scalar_t, bool upper>
+__global__
+void triu_tril_kernel(
+    scalar_t* result, scalar_t* self, int64_t k, int64_t N,
```
cuda kernels look simpler now, don't they? :+1:

vishwakftw replied:
Of course, they do. Thank you!

ngimel commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -396,4 +388,136 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  return result;
+}
+
+Tensor& triu_cpu_(Tensor &self, int64_t k) {
```
Is it possible to use the same pattern as on the gpu side here, where triu_gpu_ just calls triu_gpu_out?

vishwakftw replied:
In the CPU implementation, we save time by preventing unnecessary assignments, hence this approach. I too thought about de-duplicating the CPU implementation, but this is what prevented me from doing so.

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -461,6 +461,81 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+}
+
+Tensor& triu_cuda_(Tensor &self, int64_t k) {
+  if (!checkBatchContiguous(self)) self = self.contiguous();
```
This will be done in tril_cuda_out, it's not needed here?

vishwakftw replied:
I did realize the redundancy of the check, however here we make the input contiguous inplace if the condition is not met, whereas in `triu_cuda_out`, this is not the behavior. I will try to eliminate this check.

ngimel reviewed (approved):
Did not check the cpu part, I requested small changes, but this is generally good now.

mrshenli commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -461,6 +461,81 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+          self_batch_stride = self.dim() > 2 ? self.stride(-3) : 1,
+          self_row_stride = self.stride(-2), self_col_stride = self.stride(-1);
+  dim3 dim_block = cuda::getApplyBlock();
+  dim3 dim_grid((mat_size * n_batches + dim_block.x * n_batches - 1) / (dim_block.x * n_batches), n_batches);
```
will the following code do the same?

```c++
dim_grid((mat_size + dim_block.x - 1) / dim_block.x, n_batches)
```

vishwakftw replied:
Yes, it should. I just used the same logic as done in the `cuda::getApplyGrid` function, hence this code. Thank you.

vishwakftw replied:
Thank you for the suggestion. Yes, it would do the same.

vishwakftw commented on aten/src/TH/generic/THTensorMath.h:
```diff
@@ -102,7 +102,6 @@ TH_API void THTensor_(randperm)(THTensor *r_, THGenerator *_generator, int64_t n
 TH_API void THTensor_(sort)(THTensor *rt_, THLongTensor *ri_, THTensor *t, int dimension, int descendingOrder);
 TH_API void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, int64_t k, int dim, int dir, int sorted);
-TH_API void THTensor_(tril)(THTensor *r_, THTensor *t, int64_t k);
 TH_API void THTensor_(triu)(THTensor *r_, THTensor *t, int64_t k);
```
`triu` is used for certain LAPACK/MAGMA ops, which is why I haven't removed them.

vishwakftw commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -1202,37 +1202,6 @@ void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, int64_t k, i
-  }
-}
-
 void THTensor_(triu)(THTensor *r_, THTensor *t, int64_t k)
```
Please refer to above comment.

vishwakftw commented on aten/src/THC/generic/THCTensorMath.h:
```diff
@@ -12,7 +12,6 @@ THC_API void THCTensor_(cat)(THCState *state, THCTensor *result, THCTensor *ta,
 THC_API void THCTensor_(nonzero)(THCState* state, THCudaLongTensor *tensor, THCTensor *self);
 
-THC_API void THCTensor_(tril)(THCState *state, THCTensor *self, THCTensor *src, int64_t k);
 THC_API void THCTensor_(triu)(THCState *state, THCTensor *self, THCTensor *src, int64_t k);
```
Please refer to above comment.

vishwakftw commented on aten/src/THC/generic/THCTensorMathPairwise.cu:
```diff
@@ -168,35 +168,6 @@ void THCTensor_(remainder)(THCState *state, THCTensor *self_, THCTensor *src_, s
-  THCudaCheck(cudaGetLastError());
-}
-
 void THCTensor_(triu)(THCState *state, THCTensor *self_, THCTensor *src_, int64_t k)
```
Please refer to above comment.

vishwakftw commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -41,6 +41,17 @@ static inline int64_t matrixStride(const Tensor& batched_matrices) {
 }
 
+// Checks if a given tensor has batch contiguous
+static inline bool checkBatchContiguous(const Tensor& tensor) {
```
Oh yes, my bad. I will fix this.

vishwakftw committed:
CR.5

- Make the contiguous check weaker
- Grid dimension computation simplification
- Make (batch) contiguous checks more conservative

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batched-tril-triu

mrshenli commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -396,4 +388,146 @@ Tensor& cholesky_out(Tensor &result, const Tensor &self, bool upper) {
+  // Else, we are happy with contiguous tensors as is, and can safely bypass the check
+  // This check is written in this way, because if the first predicate fails, then
+  // the next one is not checked
+  if (!self.is_contiguous() && !checkBatchContiguous(self)) self = self.contiguous();
```
nit:

1. It seems the `checkBatchContiguous` is always checked together with `is_contiguous()`, can you move `is_contiguous()` invocation into `checkBatchContiguous `?
2. As you have skipped the `dims == 3` case, `checkBatchContiguous ` is no longer checking batch contiguous (it was), because it can be non-contiguous on the `-3` dimension (please do correct me if I am wrong).  Could you please try to use a more appropriate name and add comments to explain why you skipped `dims == 3` case (e.g., `apply_triu_tril ` explicitly handles non-contiguous cases in -3 dim)? Thanks.

mrshenli reviewed (approved):
Thank you for taking care of this. Great work!

vishwakftw committed:
CR.6

- Move complete contiguity check to check function
- Rename check function

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batched-tril-triu

yongheng1991 commented:
Hi,
Thanks very much for your contribution.
Have you ever considered the problem when the batch size is larger than 65536? I got an error when I did that. I think it might because of that you used "magma_int_t" to define the 'batch_size' in https://github.com/pytorch/pytorch/blob/646cb6157df948643ec98b8947026ccaee5c7415/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu#L775
And I think the "magma_int_t" uses 32-bit integer by default.
Best

vishwakftw commented:
Hi Yongheng, thanks for the message. Are you facing issues in `triu` / `tril` or in `triangular_solve`?

yongheng1991 commented:
> Hi Yongheng, thanks for the message. Are you facing issues in `triu` / `tril` or in `triangular_solve`?

Yes. I am using it in a batch-wise eigenvalue decomposition. I have tested it has an error when the batch size is larger than 65535.
Here is a part of code:
        auto tmp_gxu =at::triu(gx.transpose(1, 2), 1);
        gx=gx.triu_().add_(tmp_gxu);  
And this is the cuda error:
RuntimeError: CUDA error: invalid configuration argument (triu_tril_cuda_template at /opt/conda/conda-bld/pytorch_1556653099582/work/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:709)
frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x45 (0x7fb0c0dd7dc5 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: at::Tensor& at::native::triu_tril_cuda_template<true>(at::Tensor&, at::Tensor const&, long, char const*) + 0x2ce (0x7fb0c67bec0e in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libcaffe2_gpu.so)
frame #2: at::native::triu_cuda_out(at::Tensor&, at::Tensor const&, long) + 0xc6 (0x7fb0c67b27a6 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libcaffe2_gpu.so)
frame #3: at::CUDAType::triu_out(at::Tensor&, at::Tensor const&, long) const + 0xd6 (0x7fb0c53fefa6 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libcaffe2_gpu.so)
frame #4: at::native::triu(at::Tensor const&, long) + 0x6a (0x7fb0c163d94a in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libcaffe2.so)
frame #5: at::TypeDefault::triu(at::Tensor const&, long) const + 0x5d (0x7fb0c1a749cd in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libcaffe2.so)
frame #6: torch::autograd::VariableType::triu(at::Tensor const&, long) const + 0x44f (0x7fb0be7a0bbf in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch.so.1)
frame #7: batch_symeig_backward(at::Tensor const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, bool) + 0x75c (0x7fb0aa68052c in /home/zhao/anaconda3/lib/python3.6/site-packages/torch_autograd_solver-0.0.0-py3.6-linux-x86_64.egg/torch_autograd_solver_aten.cpython-36m-x86_64-linux-gnu.so)
frame #8: <unknown function> + 0xeea4 (0x7fb0aa685ea4 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch_autograd_solver-0.0.0-py3.6-linux-x86_64.egg/torch_autograd_solver_aten.cpython-36m-x86_64-linux-gnu.so)
frame #9: <unknown function> + 0xf13e (0x7fb0aa68613e in /home/zhao/anaconda3/lib/python3.6/site-packages/torch_autograd_solver-0.0.0-py3.6-linux-x86_64.egg/torch_autograd_solver_aten.cpython-36m-x86_64-linux-gnu.so)
frame #10: <unknown function> + 0x13585 (0x7fb0aa68a585 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch_autograd_solver-0.0.0-py3.6-linux-x86_64.egg/torch_autograd_solver_aten.cpython-36m-x86_64-linux-gnu.so)
<omitting python frames>
frame #18: THPFunction_do_backward(THPFunction*, _object*) + 0xf6 (0x7fb0f005d8a6 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #22: torch::autograd::PyFunction::legacy_apply(std::vector<torch::autograd::Variable, std::allocator<torch::autograd::Variable> > const&) + 0xdf (0x7fb0f005dc4f in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #23: torch::autograd::PyFunction::apply(std::vector<torch::autograd::Variable, std::allocator<torch::autograd::Variable> >&&) + 0x837 (0x7fb0f005fa57 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #24: <unknown function> + 0x307622 (0x7fb0be40e622 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch.so.1)
frame #25: torch::autograd::Engine::evaluate_function(torch::autograd::FunctionTask&) + 0x385 (0x7fb0be407745 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch.so.1)
frame #26: torch::autograd::Engine::thread_main(torch::autograd::GraphTask*) + 0xc0 (0x7fb0be409740 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch.so.1)
frame #27: torch::autograd::Engine::thread_init(int) + 0x2b0 (0x7fb0be4069e0 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch.so.1)
frame #28: torch::autograd::python::PythonEngine::thread_init(int) + 0x2a (0x7fb0f0059d1a in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
frame #29: <unknown function> + 0xb8678 (0x7fb0bf4d0678 in /home/zhao/anaconda3/lib/python3.6/site-packages/torch/lib/../../../../libstdc++.so.6)
frame #30: <unknown function> + 0x76ba (0x7fb0ff7786ba in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #31: clone + 0x6d (0x7fb0ff4ae41d in /lib/x86_64-linux-gnu/libc.so.6)

vishwakftw commented:
I think I know the reason for this: the batches for the triu and tril use the y-dimension of the grid, whose limit is 65535. I will try to fix within the next week. For now, you could try mini-batching them in a loop. Sorry about the inconvenience.

yongheng1991 commented:
> I think I know the reason for this: the batches for the triu and tril use the y-dimension of the grid, whose limit is 65535. I will try to fix within the next week. For now, you could try mini-batching them in a loop. Sorry about the inconvenience.

Thanks very much. Glad to help

vishwakftw commented:
@yongheng1991 Please take a look at #21067 which adds support for batch sizes > 65535.
